### ISSUE TITLE
Rescued exceptions leave their handlers behind

### ISSUE THREAD

**Issue by nddrylliog** [fasterthanlime/rock#680]
See this source code:

``` ooc
    init: func {
        if (Env get("NO_FANCY_BACKTRACE")) {
            fancy? = false
            return
        }

        try {
            lib = Dynlib new("backtrace")
        } catch (e: DynlibException) {
            "Couldn't load at first.." println()
            lib = null
        }

        if (!lib) try {
            lib = Dynlib new("./backtrace")
        } catch (e: DynlibException) { 
            "Couldn't load at second.." println()
            lib = null
        }

        if (lib) {
            "Could load!" println()
            initFuncs()

            // get rid of rock's built-in stuff
            Env set("FANCY_BACKTRACE", "1")
        } else {
            "Could not load!" println()
            // couldn't load :(
            fancy? = false
        }
    }
```

Fun, yes? Now consider the generated C code for one of the try/catch blocks:

``` c
    {
        lang_Exception__StackFrame __frame__ = lang_Exception___pushStackFrame();
        if (lang_Exception__JmpBuf_setJmp((*(__frame__)).buf) == 0) {
            this->lib = dynlib__Dynlib_new(__strLit127);
        }
        else {
            lang_Exception__Exception* __matchExpr128 = lang_Exception___getException();
            if (lang_types__Object_instanceOf__quest((lang_types__Object*) __matchExpr128, (lang_types__Class*) (dynlib__DynlibException_class()))){
                dynlib__DynlibException* e = NULL;
                e = ((dynlib__DynlibException*) (__matchExpr128));
                lang_String__String_println(__strLit291);
                this->lib = (void*) NULL;
            } else {
                lang_Exception__Exception_rethrow(lang_Exception___getException());
            };
        }
    }
```

Alright so - we push an exception handler (via `lang_Exception___pushStackFrame`), then we setjmp (the try block) - first time goes through the if body, we attempt our risky code.

If something throws an exception, it does longjmp - and it'll end up in the else block, where.. we attempt to match the exception type. If it's one of the exception types we expect (in this case, `DynlibException`), we do some handling and go on our merry way, otherwise, we rethrow.

In this case, let's say we catch the exception, handle it ourselves, then move on. The exception handler was never unregistered. It's still there! So, what happens if something else throws an exception later that doesn't get caught by a closer exception handler?

A stack trace like this:

```
[OutOfBoundsException in ArrayList]: Trying to access an element at offset 0, but size is only 0!
[fancy backtrace]
0     print_stacktrace                             (at /home/amos/Dev/tests/backtrace-universal/backtrace.c:453)                            
1     killpg                                       (at (null):0)                                                                            
2     GI raise()                in                 (at /build/eglibc-MUWt1e/eglibc-2.17/signal/../nptl/sysdeps/unix/sysv/linux/raise.c:56)  
3     GI abort()                in                 (at /build/eglibc-MUWt1e/eglibc-2.17/stdlib/abort.c:92)                                  
4     Exception throw_impl()    in lang/Exception  (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/sdk/lang/Exception.c:284)
5     Exception throw()         in lang/Exception  (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/sdk/lang/Exception.c:316)
6     Exception rethrow_impl()  in lang/Exception  (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/sdk/lang/Exception.c:291)
7     Exception rethrow()       in lang/Exception  (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/sdk/lang/Exception.c:320)
8     BacktraceHandler init()   in obacktrace      (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/obacktrace.c:75)   
9     App run_impl()            in otest           (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/otest.c:36)        
10    App run()                 in otest           (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/otest.c:48)        
11    main                                         (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/otest.c:173)       
12    libc_start_main()         in                 (at /build/eglibc-MUWt1e/eglibc-2.17/csu/libc-start.c:294)                               
13    _start                                       (at (null):0)                 
```

But, that's clearly wrong, right? (Still hella fancy though). There's no way the OutOfBoundsException was thrown from backtrace-universal - it was thrown somewhere in our test code. Yet the stack trace shows otherwise.

What if we add a `lang_Exception___popStackFrame` ourselves at the end of the try block? Example:

``` c
    {
        lang_Exception__StackFrame __frame__ = lang_Exception___pushStackFrame();
        if (lang_Exception__JmpBuf_setJmp((*(__frame__)).buf) == 0) {
            this->lib = dynlib__Dynlib_new(__strLit127);
        }
        else {
            lang_Exception__Exception* __matchExpr128 = lang_Exception___getException();
            if (lang_types__Object_instanceOf__quest((lang_types__Object*) __matchExpr128, (lang_types__Class*) (dynlib__DynlibException_class()))){
                dynlib__DynlibException* e = NULL;
                e = ((dynlib__DynlibException*) (__matchExpr128));
                lang_String__String_println(__strLit291);
                this->lib = (void*) NULL;
            } else {
                lang_Exception__Exception_rethrow(lang_Exception___getException());
            };
        }
        lang_Exception___popStackFrame();
    }
```

In this case, things are good:
- We register a handler
- Prepare for the worst...
- Do something risky
- Either it works out and we get out of the if, unregister the handler, all good!
- Or it doesn't, we go in the else, match the right type of Exception, do something, get out of the if, the else, unregister the handler, all good!

And indeed, the new backtrace shows that our instinct is right:

```
[fancy backtrace]
0     print_stacktrace                                   (at /home/amos/Dev/tests/backtrace-universal/backtrace.c:453)                               
1     killpg                                             (at (null):0)                                                                               
2     GI raise()                   in                    (at /build/eglibc-MUWt1e/eglibc-2.17/signal/../nptl/sysdeps/unix/sysv/linux/raise.c:56)     
3     GI abort()                   in                    (at /build/eglibc-MUWt1e/eglibc-2.17/stdlib/abort.c:92)                                     
4     Exception throw_impl()       in lang/Exception     (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/sdk/lang/Exception.c:284)   
5     Exception throw()            in lang/Exception     (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/sdk/lang/Exception.c:316)   
6     ArrayList get_impl()         in structs/ArrayList  (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/sdk/structs/ArrayList.c:85) 
7     ArrayList get()              in structs/ArrayList  (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/sdk/structs/ArrayList.c:272)
8     __OP_IDX_ArrayList_Int__T()  in structs/ArrayList  (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/sdk/structs/ArrayList.c:668)
9     foo()                        in otest              (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/otest.c:156)          
10    bar()                        in otest              (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/otest.c:161)          
11    App runToo_impl()            in otest              (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/otest.c:40)           
12    App runToo()                 in otest              (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/otest.c:52)           
13    __otest_closure351()         in otest              (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/otest.c:178)          
14    __otest_closure351_thunk()   in otest              (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/otest.c:182)          
15    loop()                       in lang/Abstractions  (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/sdk/lang/Abstractions.c:40) 
16    App run_impl()               in otest              (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/otest.c:36)           
17    App run()                    in otest              (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/otest.c:48)           
18    main                                               (at /home/amos/Dev/tests/backtrace-universal/build/rock_tmp/ooc/otest/otest.c:173)          
19    libc_start_main()            in                    (at /build/eglibc-MUWt1e/eglibc-2.17/csu/libc-start.c:294)                                  
20    _start                                             (at (null):0)                                                                  
```

TL;DR @fredreichbier forgot to pop in case things go well. (That should be added by rock).
